### PR TITLE
feat(rpc): Added starknet_getTransactionStatus and removed starknet_pendingTransactions

### PR DESCRIPTION
**Pull request by antiyro**
## feat(rpc): Added starknet_getTransactionStatus and removed starknet_pendingTransactions

This PR solves the implementation of a new call: starknet_getTransactionStatus and the removal of starknet_pendingTransactions according to the v0.5.1 RPC specs described in #1291 

`starknet_getTransactionStatus`

Gets the Transaction Status, Including Mempool Status and Execution Details
    
This method retrieves the status of a specified transaction. It provides information on
whether the transaction is still in the mempool, has been executed, or dropped from the
mempool. The status includes both finality status and execution status of the
transaction.

### Arguments

* `transaction_hash` - The hash of the transaction for which the status is requested.

### Returns

* `transaction_status` - An object containing the transaction status details:
 - `finality_status`: The finality status of the transaction, indicating whether it is
   confirmed, pending, or rejected.
 - `execution_status`: The execution status of the transaction, providing details on the
   execution outcome if the transaction has been processed.

### Expected behavior

image

## Notes

- TODO: https://github.com/antiyro/madara/blob/feat/starknet_getTransactionStatus/crates/primitives/transactions/src/lib.rs#L32
- Tests: The call has been tested manually in multiple cases but official tests needs to be implemented here: [#1298](https://github.com/keep-starknet-strange/madara/issues/1298)

### CODE REVIEW

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -400,6 +401,67 @@ where
...
+    ///
+    /// ### Returns
+    ///
+    /// * `transaction_status` - An object containing the transaction status details:
+    ///   - `finality_status`: The finality status of the transaction, indicating whether it is
+    ///     confirmed, pending, or rejected.
+    ///   - `execution_status`: The execution status of the transaction, providing details on the
+    ///     execution outcome if the transaction has been processed.
+    fn get_transaction_status(&self, transaction_hash: FieldElement) -> RpcResult<TransactionStatus> {
+        let block_hash_from_db = self
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:423]: rename to `substrate_block_hash`

File: crates/client/rpc/src/lib.rs [hunk 2]
```diff
@@ -400,6 +401,67 @@ where
...
+            .backend
+            .mapping()
+            .block_hash_from_transaction_hash(H256::from(transaction_hash.to_bytes_be()))
+            .map_err(|e| {
+                error!("Failed to get transaction's substrate block hash from mapping_db: {e}");
+                StarknetRpcApiError::TxnHashNotFound
+            })?;
+
+        let substrate_block_hash = block_hash_from_db.ok_or(StarknetRpcApiError::TxnHashNotFound)?;
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:432]: chain those operations

File: crates/client/rpc/src/lib.rs [hunk 3]
```diff
@@ -400,6 +401,67 @@ where
...
+
+        let block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash)
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:434]: is it a substrate or a starknet block? Make it explicit in the variable name

File: crates/client/rpc/src/lib.rs [hunk 4]
```diff
@@ -400,6 +401,67 @@ where
...
+            .ok_or(StarknetRpcApiError::BlockNotFound)?;
+
+        let chain_id = self.chain_id()?.0.into();
+
+        let find_tx = block
+            .transactions()
+            .iter()
+            .find(|tx| tx.compute_hash::<H>(chain_id, false).0 == transaction_hash)
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:442]: don't we maintain a dbstorage mapping that spare us the need to recompute the tx hash.
I think we can retrieve it by blockhash and index.

File: crates/client/rpc/src/lib.rs [hunk 5]
```diff
@@ -400,6 +401,67 @@ where
...
+            })?;
+
+        let substrate_block_hash = block_hash_from_db.ok_or(StarknetRpcApiError::TxnHashNotFound)?;
+
+        let block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash)
+            .ok_or(StarknetRpcApiError::BlockNotFound)?;
+
+        let chain_id = self.chain_id()?.0.into();
+
+        let find_tx = block
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:439]: `found_tx` or just `tx`

File: crates/client/rpc/src/lib.rs [hunk 6]
```diff
@@ -400,6 +401,67 @@ where
...
+            .map(|tx| to_starknet_core_tx(tx.clone(), transaction_hash))
+            .ok_or(StarknetRpcApiError::TxnHashNotFound)?;
+
+        let execution_result: TransactionExecutionStatus = self
+            .client
+            .runtime_api()
+            .get_tx_execution_outcome(substrate_block_hash, Felt252Wrapper(transaction_hash).into())
+            .map_err(|e| {
+                error!(
+                    "Failed to get transaction execution outcome. Substrate block hash: {substrate_block_hash}, \
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:452]: `Failed to get transaction execution outcome through StarknetRuntimeApi. ...`

File: crates/client/rpc/src/lib.rs [hunk 7]
```diff
@@ -400,6 +401,67 @@ where
...
+                     transaction hash: {transaction_hash}, error: {e}"
+                );
+                StarknetRpcApiError::InternalServerError
+            })?
+            .map_or(TransactionExecutionStatus::Succeeded, |message| TransactionExecutionStatus::Reverted);
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:457]: `message` is not used? We discard the error message? 
Is it the expected behaviour of the RPC?

File: crates/client/rpc/src/lib.rs [hunk 8]
```diff
@@ -400,6 +401,67 @@ where
...
+
+        Ok(TransactionStatus {
+            finality_status: TransactionFinalityStatus::AcceptedOnL2,
+            execution_status: execution_result,
+        })
+    }
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:463]: What about the other finality status?
Recieved (in the memepool)
Rejected (could not be put in the mempool)

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs:442]: My bad I forgot about it, will correct it inspired on `get_transaction_by_hash`

File: crates/client/rpc/src/lib.rs (hunk 1 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs:423]: and `get_transaction_receipt`

File: crates/client/rpc/src/lib.rs (hunk 3 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs:434]: Updated it in all other rpc calls for concensus

File: crates/client/rpc/src/lib.rs (hunk 7 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs:457]: I used ExecutionResult back then but I confirm that message should not be used: `The execution status of the transaction, providing details on the execution outcome if the transaction has been processed`

File: crates/client/rpc/src/lib.rs (hunk 5 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs:439]: I used `starknet_tx` for concensus wdyt

File: crates/client/rpc/src/lib.rs (hunk 8 above)
Comment (tdelabro) [reply crates/client/rpc/src/lib.rs:475]: Okay, but create the required issue with enough background for somebody to jump in

File: crates/client/rpc-core/Cargo.toml [hunk 9]
```diff
@@ -32,6 +32,7 @@ jsonrpsee = { workspace = true, features = [
 ], default-features = true }
 mp-block = { workspace = true }
 mp-digest-log = { workspace = true }
+mp-transactions = { workspace = true, features = ["serde"] }
```
Comment (tdelabro) [inline crates/client/rpc-core/Cargo.toml:35]: if you do that `serde = ["mp-transactions/serde"]`you don't need to do this `features = ["serde"]`

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/rpc-core/Cargo.toml (hunk 9 above)
Comment (tdelabro) [reply crates/client/rpc-core/Cargo.toml:35]: But it look like serde is always imported. So its not an optional feature. So your most likely want `features = ["serde"]` and not `serde = ["mp-transactions/serde"]`

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (tdelabro) [reply crates/client/rpc/src/lib.rs]: you forgot this one

File: crates/client/rpc/src/lib.rs (hunk 6 above)
Comment (tdelabro) [reply crates/client/rpc/src/lib.rs]: And this one too

File: crates/client/rpc-core/Cargo.toml (hunk 9 above)
Comment (antiyro) [reply crates/client/rpc-core/Cargo.toml:35]: Done!

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs]: I did it in last commits

File: crates/client/rpc/src/lib.rs (hunk 6 above)
Comment (antiyro) [reply crates/client/rpc/src/lib.rs]: Same as above conversation

**Review by tdelabro — APPROVED** [review]

**Pull request merged**